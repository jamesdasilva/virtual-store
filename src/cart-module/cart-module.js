import axios from 'axios'

const INSERT_PRODUCT_IN_CART = 'leroy-merlin/cart/INSERT_PRODUCT_IN_CART'
const DELETE_PRODUCT_CART = 'leroy-merlin/cart/DELETE_PRODUCT_CART'
const SHOW_CART = 'leroy-merlin/cart/SHOW_CART'
const HIDDEN_CART = 'leroy-merlin/cart/HIDDEN_CART'
const TOGGLE_CART = 'leroy-merlin/cart/TOGGLE_CART'
const CHANGE_AMOUNT = 'leroy-merlin/cart/CHANGE_AMOUNT'
const INCREMENT_AMOUNT = 'leroy-merlin/cart/INCREMENT_AMOUNT'
const DECREASE_AMOUNT = 'leroy-merlin/cart/DECREASE_AMOUNT'
const SET_FREIGHT = 'leroy-merlin/cart/SET_FREIGHT'
const SET_CEP = 'leroy-merlin/cart/SET_CEP'

const initialState = {
  items: [],
  displayCart: false,
  freight: {
    cep: undefined,
    value: undefined
  }
}

const addNewItem = (product, state) => {
  const index = state.items.findIndex(item => item.product.id === product.id)
  if (index >= 0) {
    state.items[index].amount = state.items[index].amount + 1
    return [...state.items]
  }
  return [...state.items, { amount: 1, product: product }]
}

const changeAmount = (newAmount, id, state) => {
  const index = state.items.findIndex(item => item.product.id === id)
  if (index >= 0) {
    state.items[index].amount = parseInt(newAmount) || 0
  }
  return [...state.items]
}

const incrementAmount = (id, state) => {
  const index = state.items.findIndex(item => item.product.id === id)
  if (index >= 0) {
    state.items[index].amount = state.items[index].amount + 1
  }
  return [...state.items]
}

const decreaseAmount = (id, state) => {
  const index = state.items.findIndex(item => item.product.id === id)
  if (index >= 0 && state.items[index].amount > 0) {
    state.items[index].amount = state.items[index].amount - 1
  }
  return [...state.items]
}

const deleteProduct = (id, state) => {
  const index = state.items.findIndex(item => item.product.id === id)
  if (index >= 0) {
    state.items.splice(index, 1)
  }
  return [...state.items]
}

export default function (state = initialState, action) {
  const product = action.payload
  switch (action.type) {
    case INSERT_PRODUCT_IN_CART:
      return {
        ...state,
        items: addNewItem(product, state)
      }
    case CHANGE_AMOUNT:
      return {
        ...state,
        items: changeAmount(action.payload.newAmount, action.payload.id, state)
      }
    case INCREMENT_AMOUNT:
      return {
        ...state,
        items: incrementAmount(action.payload.id, state)
      }
    case DECREASE_AMOUNT:
      return {
        ...state,
        items: decreaseAmount(action.payload.id, state)
      }
    case DELETE_PRODUCT_CART:
      return {
        ...state,
        items: deleteProduct(action.payload.id, state)
      }
    case SHOW_CART:
      return {
        ...state,
        displayCart: true
      }
    case HIDDEN_CART:
      return {
        ...state,
        displayCart: false
      }
    case TOGGLE_CART:
      return {
        ...state,
        displayCart: !state.displayCart
      }
    case SET_CEP:
      return {
        ...state,
        freight: {
          ...state.freight,
          cep: action.payload.cep
        }
      }
    case SET_FREIGHT:
      return {
        ...state,
        freight: {
          cep: action.payload.cep,
          value: action.payload.freight
        }
      }
    default:
      return state
  }
}

export const insertProductInCart = (product) => {
  return {
    type: INSERT_PRODUCT_IN_CART,
    payload: product
  }
}

export const changeAmountOfItem = (newAmount, id) => {
  return {
    type: CHANGE_AMOUNT,
    payload: {
      newAmount,
      id
    }
  }
}

export const incrementAmountOfItem = (id) => {
  return {
    type: INCREMENT_AMOUNT,
    payload: {
      id
    }
  }
}

export const decreaseAmountOfItem = (id) => {
  return {
    type: DECREASE_AMOUNT,
    payload: {
      id
    }
  }
}

export const deleteProductOfCart = (id) => {
  return {
    type: DELETE_PRODUCT_CART,
    payload: {
      id
    }
  }
}

export const showCart = () => {
  return {
    type: SHOW_CART
  }
}

export const hiddenCart = () => {
  return {
    type: HIDDEN_CART
  }
}

export const toggleCart = () => {
  return {
    type: TOGGLE_CART
  }
}

export const setCep = (cep) => {
  return {
    type: SET_CEP,
    payload: {
      cep
    }
  }
}

export const setFreight = (cep, freight) => {
  return {
    type: SET_FREIGHT,
    payload: {
      cep,
      freight
    }
  }
}

export const getFreight = (cep) => {
  return (dispatch) => {
    const url = `https://us-central1-boitata-233919.cloudfunctions.net/api/freight/${cep}`
    axios({
      method: 'get',
      url: url,
      headers: {
        'Content-Type': 'application/json'
      }
    }).then((response) => {
      dispatch(setFreight(cep, response.data.freight))
    }).catch(() => {
      console.log('deu erro!')
    })
  }
}
