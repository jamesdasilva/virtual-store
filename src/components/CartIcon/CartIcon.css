.cart-icon {
  position: relative;
  margin-top: 14px;
  cursor: pointer;
}
.cart-icon__count {
  position: absolute;
  bottom: 12px;
  right: -10px;
  width: 20px;
  height: 20px;
  background-color: #f00;
  color: #fff;
  border-radius: 10px;
  text-align: center;
  line-height: 20px;
  font-size: 13px;
  font-family: Helvetica, Arial, sans-serif
}
.cart-icon__black-cart {
  display: block;
}
.cart-icon__white-cart {
  display: none;
}
@media screen and (max-width: 320px) {
  .cart-icon__black-cart {
    display: none;
  }
  .cart-icon__white-cart {
    display: block;
  }
}
@media screen and (min-width: 321px) and (max-width: 440px) {
  .cart-icon__black-cart {
    display: none;
  }
  .cart-icon__white-cart {
    display: block;
  }
}