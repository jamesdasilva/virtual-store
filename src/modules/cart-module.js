const INSERT_PRODUCT_IN_CART = 'leroy-merlin/cart/INSERT_PRODUCT_IN_CART'
const DELETE_PRODUCT_CART = 'leroy-merlin/cart/DELETE_PRODUCT_CART'
const SHOW_CART = 'leroy-merlin/cart/SHOW_CART'
const HIDDEN_CART = 'leroy-merlin/cart/HIDDEN_CART'
const TOGGLE_CART = 'leroy-merlin/cart/TOGGLE_CART'
const CHANGE_AMOUNT = 'leroy-merlin/cart/CHANGE_AMOUNT'
const INCREMENT_AMOUNT = 'leroy-merlin/cart/INCREMENT_AMOUNT'
const DECREASE_AMOUNT = 'leroy-merlin/cart/DECREASE_AMOUNT'

const initialState = {
  items: [],
  displayCart: false
}

const addNewItem = (product, state) => {
  const index = state.items.findIndex(item => item.product.id === product.id)
  if (index >= 0) {
    state.items[index].amount = state.items[index].amount + 1
    return [...state.items]
  }
  return [...state.items, { amount: 1, product: product }]
}

const changeAmount = (newAmount, id, state) => {
  const index = state.items.findIndex(item => item.product.id === id)
  if (index >= 0) {
    state.items[index].amount = parseInt(newAmount) || 0
  }
  return [...state.items]
}

const incrementAmount = (id, state) => {
  const index = state.items.findIndex(item => item.product.id === id)
  if (index >= 0) {
    state.items[index].amount = state.items[index].amount + 1
  }
  return [...state.items]
}

const decreaseAmount = (id, state) => {
  const index = state.items.findIndex(item => item.product.id === id)
  if (index >= 0) {
    state.items[index].amount = state.items[index].amount - 1
  }
  return [...state.items]
}

export default function (state = initialState, action) {
  const product = action.payload
  switch (action.type) {
    case INSERT_PRODUCT_IN_CART:
      return {
        ...state,
        items: addNewItem(product, state)
      }
    case CHANGE_AMOUNT:
      return {
        ...state,
        items: changeAmount(action.payload.newAmount, action.payload.id, state)
      }
    case INCREMENT_AMOUNT:
      return {
        ...state,
        items: incrementAmount(action.payload.id, state)
      }
    case DECREASE_AMOUNT:
      return {
        ...state,
        items: decreaseAmount(action.payload.id, state)
      }
    case DELETE_PRODUCT_CART:
      return {
        ...state,
        items: [...state.items]
      }
    case SHOW_CART:
      return {
        ...state,
        displayCart: true
      }
    case HIDDEN_CART:
      return {
        ...state,
        displayCart: false
      }
    case TOGGLE_CART:
      return {
        ...state,
        displayCart: !state.displayCart
      }
    default:
      return state
  }
}

export const insertProductInCart = (product) => {
  return {
    type: INSERT_PRODUCT_IN_CART,
    payload: product
  }
}

export const changeAmountOfItem = (newAmount, id) => {
  return {
    type: CHANGE_AMOUNT,
    payload: {
      newAmount,
      id
    }
  }
}

export const incrementAmountOfItem = (id) => {
  return {
    type: INCREMENT_AMOUNT,
    payload: {
      id
    }
  }
}

export const decreaseAmountOfItem = (id) => {
  return {
    type: DECREASE_AMOUNT,
    payload: {
      id
    }
  }
}

export const showCart = () => {
  return {
    type: SHOW_CART
  }
}

export const hiddenCart = () => {
  return {
    type: HIDDEN_CART
  }
}

export const toggleCart = () => {
  return {
    type: TOGGLE_CART
  }
}
