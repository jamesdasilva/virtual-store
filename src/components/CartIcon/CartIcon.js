import React from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import './CartIcon.css'

import blackCart from '../../images/cart.svg'
import whiteCart from '../../images/cart-white.svg'

import { toggleCart } from '../../modules/cart-module'

const totalItems = (items) => {
  if (items.length > 0) {
    return items.map(item => item.amount)
      .reduce((acc, value) => acc + value)
  }
  return 0
}

const CartIcon = (props) => {
  return (
    <div className='cart-icon' onClick={props.toggleCart}>
      <div className='cart-icon__count'>{totalItems(props.cart.items)}</div>
      <img className='cart-icon__black-cart' src={blackCart} width={34} height={34} />
      <img className='cart-icon__white-cart' src={whiteCart} width={34} height={34} />
    </div>
  )
}

CartIcon.propTypes = {
  toggleCart: PropTypes.func,
  cart: PropTypes.object
}

function mapStateToProps (state) {
  return {
    cart: state.cart
  }
}

function mapDispatchToProps (dispatch) {
  return bindActionCreators({ toggleCart }, dispatch)
}

export default connect(mapStateToProps, mapDispatchToProps)(CartIcon)
