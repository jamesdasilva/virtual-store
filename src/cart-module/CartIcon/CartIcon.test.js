import '@testing-library/jest-dom'
import { render, cleanup, waitForElement, fireEvent } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import CartIcon from './CartIcon'
import { toggleCart } from '../cart-module'

let mockStore;

afterEach(cleanup)

beforeEach(() => {
  mockStore = configureStore([]);
});

describe('<CartIcon /> spec', () => {
  test('Snapshot test', async () => {
    const store = mockStore({
      cart: {
        items: []
      }
    });
    const { getByTestId } = render(
      <Provider store={store}>
        <CartIcon />
      </Provider>
    )
    const cartIcon = await waitForElement(() => getByTestId('cart-icon'))
    expect(cartIcon).toMatchSnapshot()
  })
  
  test('Should display number 0 if cart is empty', async () => {
    const store = mockStore({
      cart: {
        items: []
      }
    });
    const { getByTestId } = render(
      <Provider store={store}>
        <CartIcon />
      </Provider>
    )
    const cartIconCount = await waitForElement(() => getByTestId('cart-icon-count'))
    expect(cartIconCount).toHaveTextContent('0')
  })

  test('Should display number 3 if cart has 3 items', async () => {
    const store = mockStore({
      cart: {
        items: [
          {amount: 1, product: {}},
          {amount: 2, product: {}}
        ]
      }
    });
    const { getByTestId } = render(
      <Provider store={store}>
        <CartIcon />
      </Provider>
    )
    const cartIconCount = await waitForElement(() => getByTestId('cart-icon-count'))
    expect(cartIconCount).toHaveTextContent('3')
  })

  test('Should call toggleCart function when clicked', async () => {
    const store = mockStore({
      cart: {
        items: [ ]
      }
    });
    store.dispatch = jest.fn();
    const { getByTestId } = render(
      <Provider store={store}>
        <CartIcon />
      </Provider>
    )
    const cartIcon = await waitForElement(() => getByTestId('cart-icon'))
    fireEvent.click(cartIcon)
    expect(store.dispatch).toBeCalledWith(toggleCart())
  })

})
