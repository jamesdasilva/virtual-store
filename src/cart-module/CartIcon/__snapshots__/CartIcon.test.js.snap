// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartIcon /> spec Snapshot test 1`] = `
<div
  class="cart-icon"
  data-testid="cart-icon"
>
  <div
    class="cart-icon__count"
    data-testid="cart-icon-count"
  >
    0
  </div>
  <img
    class="cart-icon__black-cart"
    height="34"
    src="test-file-stub"
    width="34"
  />
  <img
    class="cart-icon__white-cart"
    height="34"
    src="test-file-stub"
    width="34"
  />
</div>
`;
